openoa/utils/downloader.py

```diff
@@ -0,0 +1,425 @@
+
+            logger.info("Saved to: " + str(outfile) + "\n")
+
+            if outfile.endswith(".zip"):
+                with ZipFile(outfile) as zipfile:
+                    zipfile.extractall(outfile_path)
+
+
+
+def get_era5(
+        lat,
+        lon,
+        save_pathname,
+        save_filename,
+        ):
+    """
+    Get ERA5 data directly from the CDS service
+    This requires registration on the CDS service
+    See: https://cds.climate.copernicus.eu/api-how-to
+
+    Monthly 10m height data is demonstrated here, as hourly data takes too long to download, 
+    but could be amended and other CDS datasets also used (e.g. CERRA for Europe)
+
+    Args:
+        lat(:obj:`float`): Latitude in WGS 84 spatial reference system (decimal degrees)
+        lon(:obj:`float`): Longitude in WGS 84 spatial reference system (decimal degrees)
+        save_pathname(:obj:`string`): The path where the downloaded reanalysis data will be saved
+        save_filename(:obj:`string`): The file name used to save the downloaded reanalysis data
+
+    Returns:
+        df(:obj:`dataframe`): A dataframe containing time series of the requested reanalysis variables
+        Saved NetCDF annual ERA5 files
+        Saved ERA5 csv file
+    """
+
+    logger.info("Please note access to ERA5 data requires registration")
+    logger.info("Please see: https://cds.climate.copernicus.eu/api-how-to")
+
+    # set up cds-api client
+    try:
+        c = cdsapi.Client()
+    except Exception as e:
+        logger.error('Failed to make connection to cds: '+ str(e))
+        logger.error('Please see: https://cds.climate.copernicus.eu/api-how-to')
+        raise NameError(e)
+
+    # create save_pathname if it does not exist
+    if not os.path.exists(save_pathname):
+        os.makedirs(save_pathname)
+
+    # downloads all years from today back to the year 2000
+    now = datetime.datetime.now()
+    years = list(range(2000,now.year+1,1))
+
+    # get the data for the closest 9 nodes to the coordinates
+    node_spacing = 0.250500001*1
+
+    # download the data
+    for year in years:
+
+        outfile = save_pathname+save_filename+"_"+str(year)+".nc"
+
+        if year == now.year:
+            months = list(range(1,now.month,1))
+        else:
+            months = list(range(1,12+1,1))
+
+        # See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means?tab=form
+        # for formulating other requests from cds
+        if not os.path.exists(outfile) or year==now.year:
+
+            logger.info("Downloading ERA5 :"+outfile)
+
+            try:
+                c.retrieve(
+                    "reanalysis-era5-single-levels-monthly-means",
+                    {
+                        "product_type": "reanalysis",
+                        "format": "netcdf",
+                        "variable": [
+                            "10m_wind_speed", "2m_temperature", "surface_pressure",
```

Comment:
It seems like based on much of the examples and potential uses cases within OpenOA, we might want the following additional parameters. Or is there a reason that these are excluded? (Same goes for MERRA2)
- 10m u-windspeed
- 10m v-windspeed
- 100m windspeed
- 100m u-windspeed
- 100m v-windspeed

Reply:
Just a note that all variables aren't available at monthly resolution, so that's the reason for the limitation. And data are only available at 10m.

Reply:
When averaged over a month I'm not sure how useful the u and v compontent wind speeds. They lead to very misleading wind speeds (as the vectors naturally tend toward zero with changing hourly wind directions) and I'm sceptical of the value of the resulting wind direction too. So I'll leave as is at the moment. Happy to be persuaded otherwise though! 

Personally I think CDS and GES DISC need to include the 50m/100m wind speeds and wind directions in their monthly datasets, rather than just wind speed at surface level.

Reply:
This all makes sense, thanks for the additional background information for what is included here!